File: Module - Maps/Maps2/app/src/main/java/com/javadude/maps/MapsActivity.java
package com.javadude.maps;

import android.Manifest;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.location.Location;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.renderscript.Double2;
import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.appcompat.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.maps.android.PolyUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class MapsActivity extends AppCompatActivity implements OnMapReadyCallback {

	private GoogleMap mMap;
	private PermissionManager permissionManager = new PermissionManager();
	private SupportMapFragment mapFragment;
	private BitmapDescriptor blueCar;
	private LatLng savedPosition;
	private Marker marker;
	private Polyline route;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_maps);
		// Obtain the SupportMapFragment and get notified when the map is ready to be used.
		mapFragment = (SupportMapFragment) getSupportFragmentManager()
				.findFragmentById(R.id.map);

		blueCar = BitmapDescriptorFactory.fromResource(R.mipmap.car_blue);
	}

	/**
	 * Manipulates the map once available.
	 * This callback is triggered when the map is ready to be used.
	 * This is where we can add markers or lines, add listeners or move the camera. In this case,
	 * we just add a marker near Sydney, Australia.
	 * If Google Play services is not installed on the device, the user will be prompted to install
	 * it inside the SupportMapFragment. This method will only be triggered once the user has
	 * installed Google Play services and returned to the app.
	 */
	@Override
	public void onMapReady(GoogleMap googleMap) {
		mMap = googleMap;

		SharedPreferences preferences = getSharedPreferences("position", MODE_PRIVATE);
		String latitudeString = preferences.getString("latitude", null);
		String longitudeString = preferences.getString("longitude", null);
		if (latitudeString != null && longitudeString != null) {
			savedPosition = new LatLng(
					Double.parseDouble(latitudeString),
					Double.parseDouble(longitudeString)
			);
			if (marker != null)
				marker.remove();
			marker = mMap.addMarker(new MarkerOptions()
					.anchor(0.5f, 0.5f)
					.icon(blueCar)
					.title("Saved Location")
					.snippet("Lat/Lng: " + savedPosition.latitude + ", " + savedPosition.longitude)
					.position(savedPosition)
			);
		}
		permissionManager.run(enableMyLocationAction);
	}

	@Override
	protected void onPause() {
		super.onPause();
		if (mMap.isMyLocationEnabled())
			//noinspection MissingPermission
			mMap.setMyLocationEnabled(false);
		if (savedPosition != null) {
			getSharedPreferences("position", MODE_PRIVATE)
					.edit()
					.putString("latitude", savedPosition.latitude + "")
					.putString("longitude", savedPosition.longitude + "")
					.apply();
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		mapFragment.getMapAsync(this);
	}

	private PermissionAction enableMyLocationAction = new PermissionAction(this, null, Manifest.permission.ACCESS_FINE_LOCATION) {
		@Override
		protected void onPermissionGranted() {
			//noinspection MissingPermission
			mMap.setMyLocationEnabled(true);
		}

		@Override
		protected void onPermissionDenied() {
		}
	};


	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		permissionManager.handleRequestPermissionsResult(requestCode, permissions, grantResults);
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.menu_main, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch(item.getItemId()) {
			case R.id.action_navigate: {
				Uri uri = Uri.parse("google.navigation:q=" + savedPosition.latitude + "," + savedPosition.longitude);
				startActivity(new Intent(Intent.ACTION_VIEW, uri));
				return true;
			}
			case R.id.action_show_route: {
				Location myLocation = mMap.getMyLocation();
				new RouteFetcher().execute(myLocation.getLatitude(), myLocation.getLongitude(), savedPosition.latitude, savedPosition.longitude);
				return true;
			}
			case R.id.action_remember_location: {
				Location myLocation = mMap.getMyLocation();
				savedPosition = new LatLng(myLocation.getLatitude(), myLocation.getLongitude());
				if (marker != null)
					marker.remove();
				marker = mMap.addMarker(new MarkerOptions()
						.anchor(0.5f, 0.5f)
						.icon(blueCar)
						.title("Saved Location")
						.position(savedPosition)
				);
				return true;
			}
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	private class RouteFetcher extends AsyncTask<Double, Void, List<LatLng>> {

		@Override
		protected void onPreExecute() {
			super.onPreExecute();
			if (route != null)
				route.remove();
		}

		@Override
		protected void onPostExecute(List<LatLng> latLngs) {
			super.onPostExecute(latLngs);
			route = mMap.addPolyline(new PolylineOptions()
					.addAll(latLngs)
					.color(Color.RED)
					.width(8)
			);
		}

		@Override
		protected List<LatLng> doInBackground(Double... params) {
			double myLat = params[0];
			double myLon = params[1];
			double savedLat = params[2];
			double savedLon = params[3];
			try {
				String uriString = "http://maps.googleapis.com/maps/api/directions/json?mode=walking&origin=" + myLat + ',' + myLon + "&destination=" + savedLat + ',' + savedLon + "&sensor=true";

				URL url = new URL(uriString);
				HttpURLConnection connection = (HttpURLConnection) url.openConnection();

				connection.setRequestMethod("GET");
				connection.connect();
				int responseCode = connection.getResponseCode();
				String content = "";
				if (responseCode < 300) {
					InputStream in = connection.getInputStream();
					InputStreamReader isr = new InputStreamReader(in);
					BufferedReader br = new BufferedReader(isr);
					String line;
					while((line = br.readLine()) != null) {
						content += line + "\n";
					}
					// SHOW THE ROUTE ON THE MAP
					JSONObject object = new JSONObject(content);
					JSONArray steps = object.getJSONArray("routes").getJSONObject(0).getJSONArray("legs").getJSONObject(0).getJSONArray("steps");
					List<LatLng> allLatLngs = new ArrayList<>();
					for(int i = 0; i < steps.length(); i++) {
						JSONObject step = steps.getJSONObject(i);
						String points = step.getJSONObject("polyline").getString("points");
						List<LatLng> latLngs = PolyUtil.decode(points);
						allLatLngs.addAll(latLngs);
					}
					return allLatLngs;
				}

				throw new RuntimeException("Could not access routing information. Response from server: " + responseCode);

			} catch (IOException| JSONException e) {
				throw new RuntimeException(e);
			}
		}
	}

}

File: Module - Maps/Maps2/app/src/main/java/com/javadude/maps/MapsActivity.kt
package com.javadude.maps

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.location.Location
import android.net.Uri
import android.os.Bundle
import android.os.Looper
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptor
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.livinglifetechway.quickpermissions_kotlin.runWithPermissions


class MapsActivity : AppCompatActivity(), OnMapReadyCallback {
    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var mapFragment: SupportMapFragment
    private lateinit var blueCar: BitmapDescriptor
    private lateinit var myLocationIcon: BitmapDescriptor
    private val locationRequest = LocationRequest.create().apply {
        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        interval = 4000
        fastestInterval = 2000
    }

    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult?) {
            currentLocation = result?.locations?.last()?.toLatLng()
        }
    }

    private fun Location.toLatLng() = LatLng(latitude, longitude)

    private var currentLocation: LatLng? = null
        set(value) {
            field = value
            map?.let {
                myLocationMarker?.remove()
                value?.let { newLocation ->
                    myLocationMarker = it.addMarker(
                        MarkerOptions()
                            .anchor(0.5f, 0.5f)
                            .icon(myLocationIcon)
                            .position(newLocation)
                    )
                }
            }
        }

    private var map: GoogleMap? = null
    private var savedPosition: LatLng? = null
    private var marker: Marker? = null
    private var myLocationMarker: Marker? = null

    private var latLngBoundsPadding = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_maps)

        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        latLngBoundsPadding = resources.getDimension(R.dimen.lat_lng_bounds_padding).toInt()

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment

        blueCar = BitmapDescriptorFactory.fromResource(R.mipmap.car_blue) // already a bitmap drawable

        myLocationIcon = loadBitmapDescriptor(R.drawable.my_location)
    }

    private fun startLocationUpdates() = runWithPermissions(Manifest.permission.ACCESS_FINE_LOCATION) {
        fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper())
    }
    private fun stopLocationUpdates() = runWithPermissions(Manifest.permission.ACCESS_FINE_LOCATION) {
        fusedLocationClient.removeLocationUpdates(locationCallback)
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * This is where we can add markers or lines, add listeners or move the camera. In this case,
     * we just add a marker near Sydney, Australia.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap

        val preferences = getSharedPreferences("position", Context.MODE_PRIVATE)
        val latitudeString = preferences.getString("latitude", null)
        val longitudeString = preferences.getString("longitude", null)


        var latLngBounds : LatLngBounds? = null

        fun LatLng.addToBounds() {
            latLngBounds = latLngBounds?.including(this) ?: LatLngBounds(this, this)
        }

        // start the bounds around APL
        LatLng(39.163742, -76.900235).addToBounds()
        LatLng(39.163467, -76.897381).addToBounds()
        LatLng(39.162610, -76.899500).addToBounds()

        if (latitudeString != null && longitudeString != null) {
            LatLng(latitudeString.toDouble(), longitudeString.toDouble()).let {
                savedPosition = it
                marker?.remove()
                marker = googleMap.addMarker(
                    MarkerOptions()
                        .anchor(0.5f, 0.5f)
                        .icon(blueCar)
                        .title("Saved Location")
                        .snippet("Lat/Lng: " + it.latitude + ", " + it.longitude)
                        .position(it)
                )
            }
        }

        savedPosition?.addToBounds()

        googleMap.setOnMapLoadedCallback {
            googleMap.animateCamera(CameraUpdateFactory.newLatLngBounds(latLngBounds, latLngBoundsPadding))
        }

        googleMap.setOnMapLongClickListener {
            stopLocationUpdates()
            myLocationMarker?.remove()
            currentLocation = it
        }
    }

    @SuppressLint("MissingPermission")
    override fun onPause() {
        super.onPause()
        stopLocationUpdates()
        savedPosition?.let {
            getSharedPreferences("position", Context.MODE_PRIVATE)
                .edit()
                .putString("latitude", it.latitude.toString() + "")
                .putString("longitude", it.longitude.toString() + "")
                .apply()
        }
    }

    override fun onResume() {
        super.onResume()
        startLocationUpdates()
        mapFragment.getMapAsync(this)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_navigate -> {
                currentLocation?.let { curr ->
                    savedPosition?.let { car ->
                        // see https://developers.google.com/maps/documentation/urls/guide
                        //   for details on specifying the URI - it's consistent across all platforms
                        val uri = Uri.parse("https://www.google.com/maps/dir/?api=1&origin=${curr.latitude},${curr.longitude}&destination=${car.latitude},${car.longitude}&travelmode=walking")
                        startActivity(Intent(Intent.ACTION_VIEW, uri).apply {
                            setPackage("com.google.android.apps.maps")
                        })
                    }
                }
                return true
            }

            R.id.action_remember_location -> {
                marker?.remove()
                savedPosition = currentLocation?.apply {
                    marker = map!!.addMarker(
                        MarkerOptions()
                            .anchor(0.5f, 0.5f)
                            .icon(blueCar)
                            .title("Parked Here!")
                            .position(this)
                    )
                }
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }
}

